opentracing/span.h

```diff
@@ -1,200 +1,119 @@
-#ifndef OPENTRACING_SPAN_H
-#define OPENTRACING_SPAN_H
+#ifndef INCLUDED_OPENTRACING_SPAN_H
+#define INCLUDED_OPENTRACING_SPAN_H
+
+#include <opentracing/stringref.h>
 
 #include <string>
-#include <vector>
-#include <sstream>
 #include <stdint.h>
 
-namespace opentracing
-{
-
-class Tracer;
-class LogData;
+namespace opentracing {
 
-/**
- * A `Span` represents an active, un-finished span in the OpenTracing system.
- * Spans are created by the `Tracer` interface.
- */
-class Span
-{
-public:
-	/**
-	 * Destroys this Span.
-	 */
+class Span {
+  public:
     virtual ~Span();
 
-    /**
-     * Sets or changes the operation name.
-     */
-    virtual Span * setOperationName(const std::string & operationName);
-    
-    /**
-     * Adds a tag with any value to the `Span`. If there is a
-     * existing tag set for `key`, it is overwritten.
-     */
+    virtual int setOperation(const StringRef& operation) = 0;
+
+    virtual int finish()                   = 0;
+    virtual int finish(const uint64_t tsp) = 0;
+
+    virtual int tag(const StringRef& key, int16_t) = 0;
+    virtual int tag(const StringRef& key, int32_t) = 0;
+    virtual int tag(const StringRef& key, int64_t) = 0;
+
+    virtual int tag(const StringRef& key, uint16_t) = 0;
+    virtual int tag(const StringRef& key, uint32_t) = 0;
+    virtual int tag(const StringRef& key, uint64_t) = 0;
+
+    virtual int tag(const StringRef& key, signed char)   = 0;
+    virtual int tag(const StringRef& key, unsigned char) = 0;
+    virtual int tag(const StringRef& key, char)          = 0;
+
+    virtual int tag(const StringRef& key, bool)   = 0;
+    virtual int tag(const StringRef& key, float)  = 0;
+    virtual int tag(const StringRef& key, double) = 0;
+
+    template <typename T>
+    int tag(const StringRef& key, const T& val);
+
+    virtual int log(const StringRef& key, int16_t) = 0;
```

Comment:
re the various `log` methods... maybe we could do something like the zero-allocation approach in `opentracing-go`? in any case, it's important for `log` to support multiple key:value fields per log call.